gnucash/import-export/import-main-matcher.c

```diff
@@ -1435,38 +1495,149 @@ void gnc_gen_trans_list_add_trans_with_ref_id (GNCImportMainMatcher *gui, Transa
     GNCImportTransInfo * transaction_info = NULL;
     GtkTreeModel *model;
     GtkTreeIter iter;
-    GNCImportMatchInfo *selected_match;
-    gboolean match_selected_manually;
     g_assert (gui);
     g_assert (trans);
-
-
-    if (gnc_import_exists_online_id (trans))
+  
+    if (gnc_import_exists_online_id (trans, gui->id_hash))
         return;
     else
     {
         transaction_info = gnc_import_TransInfo_new (trans, NULL);
         gnc_import_TransInfo_set_ref_id (transaction_info, ref_id);
+        // It's much faster to gather the imported transactions into a GSList than directly into the
+        // treeview.
+        gui->temp_trans_list = g_slist_prepend (gui->temp_trans_list, transaction_info);
+    }
+    return;
+}
 
-        gnc_import_TransInfo_init_matches (transaction_info,
-                                           gui->user_settings);
+// This creates lists of all splits that could match one of the imported transactions based on their
+// account and date and organized per account
+static GHashTable*
+create_list_of_potential_matches (GNCImportMainMatcher *gui, GtkTreeModel* model, Query *query, gint match_date_hardlimit)
+{
+    GList* query_return = NULL;
+    Account *import_trans_account;
+    time64 import_trans_time, min_time=G_MAXINT64, max_time=0;
+    GList* all_accounts = NULL;
+    static const int secs_per_day = 86400;
+    GSList* imported_trans;
+    GList* potential_match;
+    GHashTable* lists_per_accounts = g_hash_table_new (g_direct_hash, g_direct_equal);
```

Comment:
This does not work because when I do `g_hash_table_replace (lists_per_accounts, import_trans_account, per_account_list);` at line 1563 the delete function is called on the list. We don't want the list to be destroyed, to replace it with a new list, we only want to add an element to the existing list. I can't use append instead of prepend as that's also likely to change the pointer.
I did try, and got a crash.

Reply:
Oh, right. You could change line 1563 to
```C
   g_hash_table_steal (lists_per_accounts, import_trans_account);
   g_hash_table_add (lists_per_accounts, import_trans_account, per_account_list);
```
But that's pretty opaque.  You could add a level of indirection with a struct containing only the GSList* and provide a ctor and dtor so that the hash table's value would be stable, but I think it not worth the effort.

Reply:
Right. It's annoying that you can't append to a list without having the original pointer possibly invalidated, I don't understand what that's the case (I can see why for prepend, but not for append). If that were the case, it would make the process a lot simpler, you wouldn't have to change anything to the hash table, just keep adding to the list.

Reply:
Maybe this, though, replacing 1561-1563:
```C
// Steal the GSList from the lists_per_accounts hash so that the value destructor isn't called on it.
g_hash_table_steal_extended (lists_per_accounts, import_trans_account, NULL, &per_account_list);
per_account_list = g_slist_prepend (per_account_list, potential_match->data);
g_hash_table_add (lists_per_accounts, import_trans_account, per_account_list);
```
I checked the code and `g_hash_table_steal_extended` does `NULL` the output ptrs if the key isn't found.

Reply:
Ah, nice, yes, that way you get both the initialization and the update in one shot. I'll do that.

Reply:
You wouldn't want to append anyway. It has to traverse the list to find the end so it gets really slow as the list grows. That's why the docs say that in general you should reverse, prepend, and reverse if you need to append in a loop and care about the order. That said the [returned pointer *is* stable unless the incoming is NULL](https://gitlab.gnome.org/GNOME/glib/-/blob/master/glib/gslist.c#L227).

Reply:
Ah! The doc says that the pointer might change, I guess they mean if you pass a NULL! But yes, append is quite slow for a long list.

Reply:
I think it's more likely that whoever wrote the doc comment copied the one from `g_list_prepend` and was too lazy to edit it to be more precise. Either that or they envisioned eventually creating a pre-allocated memory block to keep the list elements contiguous, which would vastly speed up traversal by improving cache usage but would also require reallocating when the list got too big for the existing block. That would require a stronger type guarantee that's difficult to pull off when you're in the habit of passing `void*` to everything so no doubt the vision died young.